File: java/client/src/main/java/glide/api/models/commands/MigrateOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands;

import glide.api.commands.GenericBaseCommands;
import java.util.ArrayList;
import java.util.List;
import lombok.Builder;

/**
 * Optional arguments for {@link GenericBaseCommands#migrate(String, long, String, long, long,
 * MigrateOptions)} command.
 *
 * @see <a href="https://valkey.io/commands/migrate/">valkey.io</a>
 */
@Builder
public final class MigrateOptions {
    /** Valkey API keyword for COPY option. */
    public static final String COPY_VALKEY_API = "COPY";

    /** Valkey API keyword for REPLACE option. */
    public static final String REPLACE_VALKEY_API = "REPLACE";

    /** Valkey API keyword for AUTH option. */
    public static final String AUTH_VALKEY_API = "AUTH";

    /** Valkey API keyword for AUTH2 option. */
    public static final String AUTH2_VALKEY_API = "AUTH2";

    /** If set, do not remove the key from the source instance. */
    private final boolean copy;

    /** If set, replace existing key on the destination instance. */
    private final boolean replace;

    /** Password for authentication to the destination instance. */
    private final String password;

    /** Username for authentication to the destination instance (requires password). */
    private final String username;

    /**
     * Converts MigrateOptions into a String[].
     *
     * @return String[]
     */
    public String[] toArgs() {
        List<String> args = new ArrayList<>();

        if (copy) {
            args.add(COPY_VALKEY_API);
        }

        if (replace) {
            args.add(REPLACE_VALKEY_API);
        }

        if (username != null && password != null) {
            args.add(AUTH2_VALKEY_API);
            args.add(username);
            args.add(password);
        } else if (password != null) {
            args.add(AUTH_VALKEY_API);
            args.add(password);
        }

        return args.toArray(new String[0]);
    }
}

File: go/options/migrate_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

import (
	"fmt"

	"github.com/valkey-io/valkey-glide/go/v2/constants"
)

// MigrateOptions represents optional arguments for the Migrate and MigrateWithOptions commands.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/migrate/
type MigrateOptions struct {
	// If true, do not remove the key from the source instance.
	Copy bool
	// If true, replace the existing key on the destination instance.
	Replace bool
	// Authentication password for the destination instance.
	Password string
	// Authentication username for the destination instance (requires Password).
	Username string
}

// NewMigrateOptions creates a new MigrateOptions with default values.
func NewMigrateOptions() *MigrateOptions {
	return &MigrateOptions{}
}

// SetCopy sets the COPY flag.
func (o *MigrateOptions) SetCopy() *MigrateOptions {
	o.Copy = true
	return o
}

// SetReplace sets the REPLACE flag.
func (o *MigrateOptions) SetReplace() *MigrateOptions {
	o.Replace = true
	return o
}

// SetPassword sets the AUTH password. Clears any previously set username.
func (o *MigrateOptions) SetPassword(password string) *MigrateOptions {
	o.Password = password
	o.Username = "" // clear username to avoid ambiguous AUTH2 state
	return o
}

// SetAuth2 sets the AUTH2 username and password.
func (o *MigrateOptions) SetAuth2(username, password string) *MigrateOptions {
	o.Username = username
	o.Password = password
	return o
}

// ToArgs converts MigrateOptions to a string slice for use in the MIGRATE command.
func (o *MigrateOptions) ToArgs() ([]string, error) {
	args := []string{}
	if o.Copy {
		args = append(args, constants.CopyKeyword)
	}
	if o.Replace {
		args = append(args, constants.ReplaceKeyword)
	}
	if o.Username != "" && o.Password == "" {
		return nil, fmt.Errorf("username requires a password")
	}
	if o.Username != "" && o.Password != "" {
		args = append(args, constants.Auth2Keyword, o.Username, o.Password)
	} else if o.Password != "" {
		args = append(args, constants.AuthKeyword, o.Password)
	}
	return args, nil
}
